uarray/_uarray_dispatch.cxx

```diff
@@ -0,0 +1,491 @@
+
+#include <Python.h>
+#include <utility>
+#include <new>
+#include <cstddef>
+
+namespace {
+class py_ref
+{
+	explicit py_ref(PyObject * object): obj(object) {}
+public:
+
+	py_ref(): obj(nullptr) {}
+	py_ref(std::nullptr_t): py_ref() {}
+
+	py_ref(const py_ref & other): obj(other.obj) { Py_XINCREF(obj); }
+	py_ref(py_ref && other): obj(other.obj) { other.obj = nullptr; }
+
+	static py_ref steal(PyObject * object) { return py_ref(object); }
+
+	static py_ref ref(PyObject * object)
+		{
+			Py_XINCREF(object);
+			return py_ref(object);
+		}
+
+	~py_ref(){ Py_XDECREF(obj); }
+
+	py_ref & operator = (const py_ref & other)
+		{
+			py_ref(other).swap(*this);
+			return *this;
+		}
+
+	py_ref & operator = (py_ref && other)
+		{
+			py_ref(std::move(other)).swap(*this);
+			return *this;
+		}
+
+	void swap(py_ref & other)
+		{
+			std::swap(other.obj, obj);
+		}
+
+	explicit operator bool () const { return obj != nullptr; }
+
+	operator PyObject* () const { return get(); }
+
+	PyObject * get() const { return obj; }
+	PyObject * release()
+		{
+			PyObject * t = obj;
+			obj = nullptr;
+			return t;
+		}
+private:
+	PyObject * obj;
+};
+
+template <typename ... Ts>
+py_ref py_make_tuple(Ts... args)
+{
+	using py_obj = PyObject *;
+	return py_ref::steal(PyTuple_Pack(sizeof...(args), py_obj{args}...));
+}
+
+
+struct py_func_args { py_ref args, kwargs; };
+
+struct Function
+{
+	PyObject_HEAD
+	py_ref extractor_, replacer_;   // functions to handle dispatchables
+	py_ref backends_;               // function returning backends as iterable
+	py_ref def_args_, def_kwargs_;  // default arguments
+	py_ref def_impl_;               // default implementation
+	py_ref dict_;                   // __dict__
+
+	PyObject * call(PyObject * args, PyObject * kwargs);
+
+	py_func_args replace_dispatchables(
+		PyObject * backend, PyObject * args, PyObject * kwargs, PyObject * coerce);
+
+	py_ref canonicalize_args(PyObject * args);
+	py_ref canonicalize_kwargs(PyObject * kwargs);
+
+	static void dealloc(Function * self)
+		{
+			PyObject_GC_UnTrack(self);
+			self->~Function();
+			Py_TYPE(self)->tp_free(self);
+		}
+
+	static PyObject * new_(PyTypeObject * type, PyObject * args, PyObject * kwargs)
+		{
+			auto self = reinterpret_cast<Function *>(type->tp_alloc(type, 0));
+			if (self == nullptr)
+				return nullptr;
+
+			// Placement new
+			self = new (self) Function;
+			return reinterpret_cast<PyObject *>(self);
+		}
+
+	static int init(Function * self, PyObject * args, PyObject * /*kwargs*/)
+		{
+			PyObject * extractor, * replacer;
+			PyObject * backends;
+			PyObject * def_args, * def_kwargs;
+			PyObject * def_impl;
+
+			if (!PyArg_ParseTuple(
+				    args, "OOOO!O!O",
+				    &extractor,
+				    &replacer,
+				    &backends,
+				    &PyTuple_Type, &def_args,
+				    &PyDict_Type, &def_kwargs,
+				    &def_impl))
+			{
+				return -1;
+			}
+
+			if (!PyCallable_Check(extractor) || !PyCallable_Check(replacer))
```

Comment:
Can we make the replacer optional as well? Only time it's needed is `__ua_convert__`.

Reply:
If there is no replacer, wouldn't that break all backends that define `__ua_convert__`?  Remembering that these would likely be defined by different people.

Reply:
I propose that the backend author would document the inability to support `__ua_convert__`